docs/coding_best_practices.md

```diff
@@ -0,0 +1,265 @@
+   components with well-known behavior, which allows one to write concise and
+   idiomatic code. It usually pays off to check if `collections`, `itertools` or
+   `functools` contain something that could be reused.
+
+ - When possible and practical implement
+   [dunder methods](https://docs.python.org/3/reference/datamodel.html#basic-customization)
+   for you classes.
+
+ - When defining a base class consider making it abstract, this would
+   disambiguate the intent. While there is some room for doubt with
+   `NotImplementedError` (maybe it would be just fine if I have implemented it
+   right here?) there is no doubt with `@abc.abstractmethod`.
+
+### Favor styles which are easier to reason about
+
+ - When possible prefer immutable objects, especially when defining containers
+   with constants.  Use `frozenset` when there are no plans to extend it, use
+   tuples instead of lists for static data.
+
+ - When possible prefer building new, modified container objects from
+   old ones to modifying existing container objects in place. A function that
+   takes an argument and constructs a return value basing on it is easier to
+   reason about than a function that modifies its argument.
+
+ - Refresh recommendations from  [PEP8](https://peps.python.org/pep-0008)
+   and [PEP20](https://peps.python.org/pep-0020);
+
+
+### Write clear tests which actually test functionality
+
+When writing tests remember, that tests look like code, but have a very
+different nature.
+
+ - Unlike for code it is ok for a test case to mostly repeat some other test.
+
+ - Unlike in actual code there could be too much abstraction in tests. If
+   understanding what and how exactly is being tested requires jumping through
+   several modules chances are the test case is overly abstracted;
+
+ - There is no limit on test names, it also makes a lot of sense to have
+   longer, more descriptive names for test cases. Consider
+   ```def test_foo()```
+   vs 
+   ```def test_foo_can_be_created_from_any_standard_source()```
+   A failure report in the first case tells a user that a test has just failed,
+   in the second case it also narrows down the scope of what could have gone
+   wrong.
+
+ - Prefer Arrange-Act-Assert pattern when possible.
+
+ - Try not to overload individual assertion statements with constructor
+   statements, especially with ones that span multiple lines. Remember, aliases
+   are cheap!
+
+ - Consider adding some meaningful message to `assert` statements.
+
+ - Please avoid extending existing test cases when providing new functionality.
+   If you find that some existing test is mostly fit for the task do not
+   hesitate to copy it over, name it unambiguously and make necessary
+   modifications to the code.
+
+ - Adding a new test parameters group, on the other hand, is a great idea!
+
+ - Consider adding a comment outlining what the test is about to do on higher
+   level when the test is complex.
+
+
+## FAQ
+
+Q. Is this guide exhaustive?
```

Comment:
![image](https://github.com/user-attachments/assets/40acdecd-30f6-410f-baee-c167381a4e91)
This doesn't render well